### PR TITLE
feat(arithmetization): sha2 precompile

### PR DESCRIPTION
**Pull request by arijitdutta67**
This PR implements issue(s) #

### Checklist

* [ ] I wrote new tests for my new core changes.
* [ ] I have successfully ran tests, style checker and build against my new changes locally.
* [ ] If this change is deployed to any environment (including Devnet), E2E test coverage exists or is included in this
  PR.
* [ ] I have informed the team of any breaking changes if there are any.

### CODE REVIEW

File: prover-ray/zkcdriver/testdata/sha2_run.zkc [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+//   in_msg         concatenated message bytes for all vectors, byte-addressed
+//   in_msg_length  byte length of each vector's message
+//   in_hash        concatenated 32-byte expected digests, byte-addressed
+// Each message is copied into guest RAM, hashed via the sha2 accelerator, and
+// the 32 output bytes are compared against the expected digest.
+//
+// This is a copy of the sha2_direct.zkc in the arithmetization folder. We add it
+// to run the local zkcdriver test covering all the edge cases in the json file
+
+include "../../../arithmetization/src/main/lib/sha2/impl.zkc"
```
[the remark this answers is not in the document]
Comment (arijitdutta67) [reply prover-ray/zkcdriver/testdata/sha2_run.zkc:16]: fixed.

File: arithmetization/src/test/Makefile [hunk 2]
```diff
@@ -326,6 +327,33 @@ keccak-zig-json: keccak-zig-build
...
+	$(ZKC) exec --check $(SHA2_ZKC_INPUT) $(SHA2_ZKC_MAIN)
+
+sha2-zig-build:
+	$(MAKE) -f $(MAKEFILE_DIR)Makefile compile \
+		SHA2_ACCEL=true \
+		TEST=sha2/sha2_provider.zig
+
+# The full interpreter's fast backend currently fails while compiling an
+# unchanged pre-existing instruction, so run this integration test in tracing mode.
+sha2-zig-exec: sha2-zig-build
```
[the remark this answers is not in the document]
Comment (arijitdutta67) [reply arithmetization/src/test/Makefile:352]: fixed.

File: arithmetization/src/test/zig/src/sha2/sha2_provider.zig [hunk 3]
```diff
@@ -0,0 +1,276 @@
+//! Self-checking, freestanding SHA-256 provider test.
```
Comment (lorenzogentile404) [inline arithmetization/src/test/zig/src/sha2/sha2_provider.zig:1]: If you look at other `*_provider.zig` files, if I remember correctly, the logic contained in the file is the essential one to connect the pieces. Here is seems there is more than that, so in case it may be worth splitting it.

File: arithmetization/src/test/zkc/sha2/sha2_direct.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (lorenzogentile404) [inline arithmetization/src/test/zkc/sha2/sha2_direct.json:1]: Is this is a file you give directly to the interpreter for testing?

File: riscv-guests/l2-execution/Makefile [hunk 5]
```diff
@@ -81,7 +85,7 @@ clean:
 	rm -rf $(BUILD_CACHE_DIR)
 
 compile:
-	$(ZIG) build $(ZIG_BUILD_FLAGS) $(ZIG_FETCH_FLAGS) $(if $(filter true,$(KECCAK_ACCEL)),-Dkeccak-accel=true)
+	$(ZIG) build $(ZIG_BUILD_FLAGS) $(ZIG_FETCH_FLAGS) $(if $(filter true,$(KECCAK_ACCEL)),-Dkeccak-accel=true) $(if $(filter true,$(SHA2_ACCEL)),-Dsha2-accel=true)
```
Comment (lorenzogentile404) [inline riscv-guests/l2-execution/Makefile:88]: Maybe here we can have an `ACCEL` flag that is generic and used for every precompile. Does it make sense to you?

**Review by lorenzogentile404 — CHANGES_REQUESTED** [review]
Left a few comments. LGTM overall.

File: arithmetization/src/test/zkc/sha2/sha2_direct.json (hunk 4 above)
Comment (arijitdutta67) [reply arithmetization/src/test/zkc/sha2/sha2_direct.json:1]: Yes. This is used to run the test command `make sha2-zkc-exec` and `make sha2-zkc-check` in the arithmetization test. There is a copy of this in the zkcdriver of the prover side to generate a wiop proof for this.

File: riscv-guests/l2-execution/Makefile (hunk 5 above)
Comment (arijitdutta67) [reply riscv-guests/l2-execution/Makefile:88]: Seems to be a reasonable suggestion. But maybe we need a hybrid approach for this, so that we can selectively turn them off or on for perf and regression test. Something like:
```
// build.zig
const accel_all    = b.option(bool, "accel", "Enable all precompile accelerators (per-flags override)") orelse false;
const keccak_accel = b.option(bool, "keccak-accel", "...") orelse accel_all;
const sha2_accel   = b.option(bool, "sha2-accel",   "...") orelse accel_all;
const bn254_add_accel = b.option(bool, "bn254-add-accel", "...") orelse accel_all;
const bn254_mul_accel = b.option(bool, "bn254-mul-accel", "...") orelse accel_all;

# Makefile
ACCEL           ?= false
KECCAK_ACCEL    ?= $(ACCEL)
SHA2_ACCEL      ?= $(ACCEL)
BN254_ADD_ACCEL ?= $(ACCEL)
BN254_MUL_ACCEL ?= $(ACCEL)
```
Can do this if it sounds ok to both of you @lorenzogentile404 @ivokub .
